src/main/java/com/google/devtools/build/lib/syntax/StringModule.java

```diff
@@ -281,25 +282,29 @@ public String strip(String self, Object charsOrNone) {
             // TODO(cparsons): This parameter should be positional-only.
             legacyNamed = true,
             doc = "The maximum number of replacements.")
-      },
-      useLocation = true)
+      })
   public String replace(
-      String self, String oldString, String newString, Object maxSplitO, Location loc)
+      String self, String oldString, String newString, Object maxSplitO)
       throws EvalException {
-    StringBuffer sb = new StringBuffer();
-    Integer maxSplit =
-        Type.INTEGER.convertOptional(
-            maxSplitO, "'maxsplit' argument of 'replace'", /*label*/ null, Integer.MAX_VALUE);
-    try {
-      Matcher m = Pattern.compile(oldString, Pattern.LITERAL).matcher(self);
-      for (int i = 0; i < maxSplit && m.find(); i++) {
-        m.appendReplacement(sb, Matcher.quoteReplacement(newString));
+    int maxSplit = Integer.MAX_VALUE;
+    if (maxSplitO != Runtime.NONE) {
+      maxSplit = Math.max(0, (Integer) maxSplitO);
+    }
+    if (oldString.equals("")) {
+      StringBuilder sb = new StringBuilder();
+      for (int i = 0; i < self.length(); i++) {
+        if (maxSplit > 0) {
+          sb.append(newString);
+          maxSplit--;
+        }
+        sb.append(self.charAt(i));
+      }
+      if (maxSplit > 0) {
+        sb.append(newString);
       }
-      m.appendTail(sb);
-    } catch (IllegalStateException e) {
-      throw new EvalException(loc, e.getMessage() + " in call to replace");
+      return sb.toString();
     }
-    return sb.toString();
+    return StringUtils.replace(self, oldString, newString, maxSplit);
```

Comment:
The replace function needed here is pretty easy to write; I don't think it warrants another dependency:

```
        int start = 0;
        for (int i = 0; i < maxSplit; i++) {
            int end = self.indexOf(old, start);
            if (end < 0) break;
            sb.append(self, start, end).append(newString);
            start = end + old.length();
        }
        sb.append(self, start, self.length());
```

Reply:
I was actually planning to use `StringUtils` in other methods too like `count` and `split` to improve their performance as well.

Reply:
To be clear: the performance improvement in this change comes not from anything clever in the StringUtils package, but from avoiding the regexp package, which has substantial overheads. By all means, avoid the regexp package in count and split too, but I don't think there's any need for an additional dependency to do that. Both functions can be implemented as a simple loop around indexOf. For example:
```
    List<String> split(String str, String sep, int max) {
        List<String> res = new ArrayList<>();
        int n = 0, i = 0;
        for (;;) {
            int j = str.indexOf(sep, i);
            if (j < 0 || ++n == max) {
                res.add(str.substring(i));
                break;
            }
            res.add(str.substring(i, j));
            i = j + sep.length();
        }
        return res;
    }
```